cmd/controller/controller.go

```diff
@@ -672,15 +672,17 @@ func (c *command) startWorker(ctx context.Context, profile string, nodeConfig *v
 	return wc.Start(ctx)
 }
 
-// If we've got CA in place we assume the node has already joined previously
+// If we've got an etcd data directory in place for embedded etcd, or a ca for
+// external or other storage types, we assume the node has already joined
+// previously.
 func (c *command) needToJoin(nodeConfig *v1beta1.ClusterConfig) bool {
+	if nodeConfig.Spec.Storage.Type == v1beta1.EtcdStorageType && !nodeConfig.Spec.Storage.Etcd.IsExternalClusterUsed() {
+		return !file.Exists(filepath.Join(c.K0sVars.EtcdDataDir, "member", "snap", "db"))
```

Comment:
Might be worth adding a comment about that file, e.g. by including  a link to https://etcd.io/docs/v3.5/learning/persistent-storage-files/#bbolt-btree-membersnapdb

Reply:
done